### ISSUE TITLE
iOS: widget for iOS 14

### ISSUE THREAD

**Issue by Dejal** [samuelclay/NewsBlur#1334]
WWDC20 has deprecated the current widget mechanism, in favor of a [SwiftUI-based non-interactive design](https://developer.apple.com/widgets/).

The current implementation is still supported for now, but is limited to the bottom of the Today view instead of the home screen on iOS 14, and will probably stop working in due course.  The widget will need to be rewritten using SwiftUI.

For NewsBlur, this shouldn't be too problematic, since it displays stories, and tapping one opens the app.  It should be feasible to reimplement that using WidgetKit. Probably not urgent; won't be able to have the widget on a home screen until this is done, but the new widget wouldn't work on iOS 13 and earlier.

**Comment by samuelclay** [comment]
Saved searches should have unread counts so we can have those as widgets. Would be incredibly valuable.

**Comment by samuelclay** [comment]
I miss having the widget be a first-class citizen.

**Comment by Dejal** [comment]
I've just been doing some SwiftUI work, so I could probably redo the widget fairly easily when next on NewsBlur work (towards the end of the month).

**Comment by Dejal** [comment]
Work on the new widget is underway. It still has a bunch of appearance tweaks and functionality to complete, but the basic thing is operational.

The old widget is still there for now, for comparison; it'll be removed once the new one is done.

I merged the Catalyst branch into the Dejal branch before committing this, so the Dejal branch can be considered the TestFlight release-ready one, and Catalyst for ongoing incomplete work.

![Simulator Screen Shot - iPhone 12 Pro Max - 2021-08-11 at 21 53 23](https://user-images.githubusercontent.com/1318849/129140615-b48d4199-8347-4447-aba7-0f529fd24975.png)

**Comment by samuelclay** [comment]
Good to know on dejal vs catalyst. I'll release dejal to TestFlight now, but I would like to hit all of the other UI bugs on TestFlight so we can get the release out. I'll go trhough now and move the remaining issues that need to be fixed before launch to critical status.

**Comment by samuelclay** [comment]
Also move author names to regular case. And once we get favicon and story image preview in there, this will be ready.

**Comment by Dejal** [comment]
Okay. Hopefully I should be able to fix all of those, and finish off the widget, in my next NewsBlur work block. I'll try to get it done before the end of the month.

**Comment by Dejal** [comment]
The feed and story images are now displayed, tapping a story opens it in the app, plus various layout tweaks.

This is in the Catalyst (my work in progress) branch, but is probably good enough to go live now. So I suggest that you try it, and if you agree, merge the Catalyst branch into the Dejal branch, and do the TestFlight from that.

(Maybe we should rename the "Dejal" branch to "TestFlight", since that's what it's used for?)

There are other improvements that could be done in the future (e.g. support for multiple widgets with different content), but that would be a lot more work, so can wait for now.

If you're satisfied with the widget as-is, please close this issue.

![Widget](https://user-images.githubusercontent.com/1318849/145649022-1da9af96-f23a-40e0-be18-d68c6c4d8303.png)

**Comment by samuelclay** [comment]
This is terrific! I updated the icon for it. 

When I installed the app and then went to install the widget, it was blank. And when I went into the Widget dialog in the app, 0 sites were selected. I hit Select All and the app crashed.

Then the widget started working. But it didn't show unread vs read, so after I tapped a story, it was still unread in the widget. While writing this up, it looks like the widget refreshed, but it should somehow acknowledge reading the story.

**Comment by Splike** [comment]
Would an option be possible to show only the title and not the story contents? As you can see in the above screenshot, space is very limited and sometimes story content is redundant, as it repeats the title. This would also make room for longer titles to span more than one line.

**Comment by Dejal** [comment]
@samuelclay I'll see if I can reproduce the crash. The widget preferences in the app should be unchanged from previous versions. Widgets refresh on a timeline managed by iOS; I do try to tell it to update when a story is read, but maybe that isn't quite right; I'll take another look. 

Also, this is in the Catalyst (future work) branch; if you feel it's good enough, feel free to merge into the Dejal (TestFlight) branch. Maybe we should rename the branches?

@Splike That is certainly possible, if Sam wants it. It might have to wait for a future update, since we're working on locking down this release.

**Comment by Dejal** [comment]
In addition to the fixes in that commit, a comment on updating on reading: it does, but with the current behavior of not marking as read when showing the first story, it looks like it isn't working. Scroll to the next one and it'll be marked as read, and the widget will update. Another reason to change that behavior. (See #1500.)

I think the widget is ready for release now, so could be merged into the Dejal (TestFlight) branch. Just waiting your word.

**Comment by samuelclay** [comment]
Tapping on a story when in the loading screen takes the user to a nowhere land.
![IMG_1829](https://user-images.githubusercontent.com/44229/151026964-e0bef34f-05a7-4606-9b63-f3eb56a11fe9.PNG)

**Comment by samuelclay** [comment]
My widget doesn't quite look like yours. iPhone SE 2.
![IMG_1828](https://user-images.githubusercontent.com/44229/151027095-beffdd15-35a0-4023-8a3b-f0681790b0f0.PNG)

**Comment by samuelclay** [comment]
The widget doesn't seem to update after I read a story or when I scroll back to the homescreen it's on. Are we being rate limited by Apple?

**Comment by samuelclay** [comment]
We can make font sizes much smaller, btw. Feels like Apple has them all at 10pt while we're at 12pt-13pt. I would prefer smaller fonts and a bit more negative space between title and subtitle as well as margins all around. We can also leave out author.

Something I learned from notifications, full title is more important than a single line of title and a single line of content. Let's make the title take as much space as it needs, and if room, show content preview. But I'm leaning towards 3 story titles on small and 6 on large, much like the NYTimes and Stocks app.

![IMG_3F208EE5E61E-1](https://user-images.githubusercontent.com/44229/151029183-6b8b46ca-cbd0-4082-8c5a-2862e49d6d6b.jpeg)

**Comment by Dejal** [comment]
Remember, just going to a story won't mark it as read, until I change that behavior; you have to scroll it or go to another story. You could be rate limited if you're doing lots of testing right now.

**Comment by samuelclay** [comment]
> Remember, just going to a story won't mark it as read, until I change that behavior; you have to scroll it or go to another story.

Sure, we're about to change that behavior, but this qualifies as a mark-read event, even in the upcoming manual mark read mode.

**Comment by samuelclay** [comment]
Wait, no, in the manual mark read mode, this won't do it still. You have to hit that button. But this is the equivalent to tapping the story title in the list.

**Comment by samuelclay** [comment]
Looks good, but can we use Whitney as the font instead of the system font?

**Comment by samuelclay** [comment]
I styled it a bit. It's nearly there, but do you think we can add the time somewhere? I have the two side by side and miss knowing relevancy.

![IMG_DEBA4BA8298D-1](https://user-images.githubusercontent.com/44229/151204607-d99d50c5-01eb-4163-a738-e7b794788bf5.jpeg)

**Comment by samuelclay** [comment]
Also, does the old widget not work anymore? Seems to take me into an infinite loop.

**Comment by samuelclay** [comment]
Tapping into stories always loads a (null) feed and spins on Finding story...

**Comment by samuelclay** [comment]
The small widget still feels cramped. Can we increase padding and tamp the font size down by 1 for just the small widget?

**Comment by samuelclay** [comment]
I would also love for the widget to respect the user's image size and position preference, as it makes for terrific screenshots.

**Comment by Dejal** [comment]
I'll see what I can do to improve it.

Not sure if I can use a custom font or not; should be able to, but unsure if the widget can access it. Will try.

For the time, I could put it to the right of the feed name (right justified). Don't want to lose the vertical height for the title.

The old widget will be removed, being redundant. Just there for now for comparison.

**Comment by Dejal** [comment]
Looking good, eh?

<img width="670" alt="Widget" src="https://user-images.githubusercontent.com/1318849/151298531-6500f294-4fe0-48b8-9dfe-afcde2cd9cd4.png">

**Comment by samuelclay** [comment]
Nice, looking great. And I agree on the datetime adding visual clutter, but NewsBlur as a whole has that slight information overload that I just love.

There's a critical bug when tapping on a story:

https://user-images.githubusercontent.com/44229/151366196-33190172-b261-4ecf-8ed0-5233339264c8.mov

Also note that I merged the dejal and catalyst branch and pushed to the new testflight branch with the build number.

Lastly, only 3 stories are showing on a 10.5" iPad. Which would be fine if the font size matched. But I'd rather stick the extra 2-3 stories in there at the right size.

![IMG_B67F9B6B5F43-1](https://user-images.githubusercontent.com/44229/151360107-fc8f7fb0-d8c8-4bd5-8382-1b5defe6ba6c.jpeg)

**Comment by Dejal** [comment]
Okay, I will commit to the `Dejal` branch now, and you can merge that into `TestFlight` when ready for builds.

I have disabled the Mac edition in the `Dejal` branch, as it is not ready for release yet. There's still a lot of work to do before it is ready.

The `Catalyst` branch can be left for future work on the Mac edition; on hold for now.

**Comment by Dejal** [comment]
I can't find a simulator for a 10.5" iPad, so can't repro that one, but it looks like it is stretching out a small widget; the large widget always includes six stories, and the small three.

**Comment by samuelclay** [comment]
It seems to me that if I were to read one of the stories and the widget refreshed before any new stories came in, then it would display fewer stories. That's my hunch, but it definitely seems like its not displaying stories found in the single page of river it fetches. Are they hidden? I have a few high volume feeds (like the Hacker News firehose) that are publisher-thumbs-down but a few title-thumbs-up. That sometimes forces me to sit through a bunch of page loads as stories slowly fill up. Try that with a feed or folder and point the widget to it.

Eventually we should enable multiple configurable widgets, possibly with the folder/feed being the only configurable element.

I've also noticed that the NewsBlur widget takes a handful of seconds longer than any of the other widgets to display. Explains why it's a blank white widget in the preview screen when first installing it. Any thoughts on why that is? Can we not access the main app's cache? Can we do a handoff when clicking stories so we don't have to fetch twice and have a fast UX?

**Comment by Dejal** [comment]
The widget fetches the stories independently of the app, since it could be running when the app isn't.

It caches the feeds, stories, feed icons, and story previews, and only loads images if not cached.

The first fetch does take a little while, which can result in a blank widget, though most of the time it is populated.

It could access the data from the main app as initial data, but it'd likely be stale, so would need to load everything anyway. Plus doing that would require duplicating all the caching code from the widget extension in the app.

There may be more optimizations that can be done in due course.

As you say, we'll likely want to support multiple widgets, with the ability to specify a folder or feed for each. That's quite a bit more complicated, so I left it as a future enhancement.
